src/httpserver.cpp

```diff
@@ -207,6 +214,17 @@ std::string RequestMethodString(HTTPRequest::RequestMethod m)
 /** HTTP request callback */
 static void http_request_cb(struct evhttp_request* req, void* arg)
 {
+    // Track requests and notify when a request is completed.
+    {
+        WITH_LOCK(g_requests_mutex, g_requests.insert(req));
+        g_requests_cv.notify_all();
```

Comment:
Did you see https://github.com/bitcoin/bitcoin/pull/26742#discussion_r1056847000 ?

Reply:
I hadn't seen that, thanks for linking. 

It's a minor concern, but I'm a bit worried that we're creating a very general contract ("`g_requests_cv` will always notify for any changes to the `g_requests`") on a global variable when it's enforced outside of `g_requests`, when in the future we might easily introduce mutations on `g_requests` without notifying the cv. A more scoped cv, such as `g_requests_cv_handled` largely mitigates that.

But, again, no big concern, and it's not likely to cause any major problems should we miss something there. As you see fit.

Reply:
FYI I think one approach to improve on this contract (and as a nice benefit, also encapsulate code a bit better - I think this improves readability) in the below diff.

(`WaitUntilEmpty()` needs improvement because it shouldn't directly be logging about the HTTP server like that. I guess we also don't need a `GlobalMutex` here.)

<details>
<summary>git diff on 55bbf04ad</summary>

```diff
diff --git a/src/httpserver.cpp b/src/httpserver.cpp
index b481ef093..90ef37bd5 100644
--- a/src/httpserver.cpp
+++ b/src/httpserver.cpp
@@ -134,6 +134,40 @@ struct HTTPPathHandler
     HTTPRequestHandler handler;
 };
 
+class RequestTracker
+{
+public:
+    typedef std::unordered_set<evhttp_request*> requests_t;
+private:
+    GlobalMutex m_requests_mutex;
+    std::condition_variable m_requests_cv;
+    requests_t m_requests GUARDED_BY(m_requests_mutex);
+public:
+    void Register(struct evhttp_request* req)
+    {
+        WITH_LOCK(m_requests_mutex, m_requests.insert(req));
+        m_requests_cv.notify_all();
+    }
+
+    size_t Remove(struct evhttp_request* req)
+    {
+        auto n{WITH_LOCK(m_requests_mutex, return m_requests.erase(req))};
+        m_requests_cv.notify_all();
+        return n;
+    }
+
+    void WaitUntilEmpty()
+    {
+        WAIT_LOCK(m_requests_mutex, lock);
+        if (!m_requests.empty()) {
+            LogPrint(BCLog::HTTP, "Waiting for %d requests to stop HTTP server\n", m_requests.size());
+        }
+        m_requests_cv.wait(lock, [this]() EXCLUSIVE_LOCKS_REQUIRED(m_requests_mutex) {
+            return m_requests.empty();
+        });
+    }
+};
+
 /** HTTP module state */
 
 //! libevent event loop
@@ -149,10 +183,8 @@ static GlobalMutex g_httppathhandlers_mutex;
 static std::vector<HTTPPathHandler> pathHandlers GUARDED_BY(g_httppathhandlers_mutex);
 //! Bound listening sockets
 static std::vector<evhttp_bound_socket *> boundSockets;
-//! Track active requests
-static GlobalMutex g_requests_mutex;
-static std::condition_variable g_requests_cv;
-static std::unordered_set<evhttp_request*> g_requests GUARDED_BY(g_requests_mutex);
+static RequestTracker g_request_tracker;
+
 
 /** Check if a network address is allowed to access the HTTP server */
 static bool ClientAllowed(const CNetAddr& netaddr)
@@ -216,12 +248,9 @@ static void http_request_cb(struct evhttp_request* req, void* arg)
 {
     // Track requests and notify when a request is completed.
     {
-        WITH_LOCK(g_requests_mutex, g_requests.insert(req));
-        g_requests_cv.notify_all();
+        g_request_tracker.Register(req);
         evhttp_request_set_on_complete_cb(req, [](struct evhttp_request* req, void*) {
-            auto n = WITH_LOCK(g_requests_mutex, return g_requests.erase(req));
-            assert(n == 1);
-            g_requests_cv.notify_all();
+            assert(g_request_tracker.Remove(req) == 1);
         }, nullptr);
     }
 
@@ -477,15 +506,7 @@ void StopHTTPServer()
         evhttp_del_accept_socket(eventHTTP, socket);
     }
     boundSockets.clear();
-    {
-        WAIT_LOCK(g_requests_mutex, lock);
-        if (!g_requests.empty()) {
-            LogPrint(BCLog::HTTP, "Waiting for %d requests to stop HTTP server\n", g_requests.size());
-        }
-        g_requests_cv.wait(lock, []() EXCLUSIVE_LOCKS_REQUIRED(g_requests_mutex) {
-            return g_requests.empty();
-        });
-    }
+    g_request_tracker.WaitUntilEmpty();
     if (eventHTTP) {
         // Schedule a callback to call evhttp_free in the event base thread, so
         // that evhttp_free does not need to be called again after the handling
```
</details>

Reply:
Sorry but it's a significant change and this PR already has a pretty long history and several ACKs for the current approach. I would suggest you open a follow-up PR with this as an additional improvement.